keps/sig-cluster-lifecycle/addons/20191013-install-addons-via-kubeadm.md

```diff
@@ -0,0 +1,450 @@
+- Create an installer library runtime with: dryRun, kubeConfig, oldInstallCfg, and newInstallCfg.
+- Prune addons from the cluster if names from the previous configuration are missing in the new configuration.
+  The library should remove addons serially in reverse order.
+- Apply all addons in the new configuration.
+  The library should respect the declared, serial order.
+- Apply the new `AddonInstallerConfiguration` to the cluster ConfigMap.
+
+Removing resources from the kubernetes API currently does not propagate any "delete reason" to processes in Pods.
+This can make it difficult (even for smart addon-operators with separately installed CustomResources) to determine when to clean up with definitive convergence. One use-case is removal of iptables rules. Another is removal of an operator + cleanup of operator-managed external resources in a single declaration.
+Uninstall Support Proposal A [not currently in scope]:
+- An uninstall kustomize dir can be optionally provided with an addon (along with `name` + `ref`) when applied.
+- When the addon is removed, either explicitly or as a consequence of prune via declaration, these manifests could be applied.
+- Any uninstall Jobs can be polled for completion.
+- If successful:
+  - All addon manifests are removed.
+  - All uninstall manifests are removed.
+
+### User Stories
+
+Users may enable this feature in `kubeadm init` and `kubeadm upgrade` using the `AddonInstaller` feature gate:
+```shell
+kubeadm init --feature-gates AddonInstaller=true
+kubeadm init --config
+```
+```yaml
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: InitConfiguration
+---
+apiVersion: kubeadm.k8s.io/v1beta2
+kind: ClusterConfiguration
+featureGates:
+  AddonInstaller: true
+```
+
+Users may invoke the phase similarly:
+```shell
+kubeadm init phase addon installer --feature-gates AddonInstaller=true
+kubeadm init phase addon installer --config
+```
+The feature gate is still expected to be enabled/true when invoking the phase directly.
+
+Users may print the default `AddonInstallerConfiguration` using `kubeadm config`:
+```shell
+kubeadm config print init-defaults --feature-gates AddonInstaller=true
+# this requires adding a feature-gates flag to this command
+```
+TODO: Discuss this alternative:
+```shell
+kubeadm config print addon-defaults
+```
+Similar U/X can be expected for `kubeadm config migrate` and `kubeadm config view`
+
+The user may extend the default kubeadm config using the `AddonInstallerConfiguration` kind.
+The API details are still being finalized. Usage may look like this:
+```yaml
+---
+apiVersion: addons.config.k8s.io/v1alpha1
+kind: AddonInstallerConfiguration
+addons:
+- name: kube-proxy
+  ref: github.com/kubernetes/kubernetes//cluster/addons/kustomize/kube-proxy/?ref=v1.17.0
+- name: coredns-operator
+  ref: github.com/kubernetes-sigs/addon-operators//kustomize/coredns-operator/?ref=v0.1.0
+- name: weavenet
```

Comment:
I think it would be nice to expose a type here and some semantics around deletion.

I bring this up because I think we need more structure here to handle the operator case. An operator will likely bring along a CRD, CR, and Operator (eg rbac/deployment). I think we should ensure that the addon installer either meets or has the flexibility to implement this part of the lifecycle. 

When an addon is removed from the this list, we ought to provide a graceful cleanup. For operators this is typically done by managing a finalizer on the CR. If we treat them all as a single package then the CR/CRD/Operator will all deleted at the same time. In that case, the CR/CRD/workload may never be removed because the operator that set the finalizer has been removed.

A sketch of this structure could look like:
```yaml
- name: coredns
  refs:
  - type: apiextensions.k8s.io/v1,Kind=CustomResourceDefinition
    url: github.com/kubernetes-sigs/addon-operators/coredns-operator/config/crd
  - type: addons.x-k8s.io/v1alpha1,Kind=CoreDNS
    manifest: -|
      apiVersion: addons.k8s.io/v1alpha1
      kind: CoreDNS
      metadata:
        name: default
        namespace: kube-system
      spec:
        channel: stable
        patches:
        - apiVersion: apps/v1
          kind: Deployment
          metadata:
            name: coredns
            namespace: kube-system
            labels:
              foo: bar
  - type: kustomize # or addons.k8s.io/Operator?
    url: github.com/kubernetes-sigs/addon-operators/kustomize/coredns-operator
```

We could delete everything but the operator, wait till the objects are fully cleaned up, then delete the operator.